### PR TITLE
[POC] Decouple icons from themes vol.2

### PR DESCRIPTION
**Pull request by victorg1991**
This is a continuation of https://github.com/liferay-frontend/liferay-portal/pull/1452

### How to test this

Since it changes an humongous amount of files, and ant all is pretty much required, then:

* Go to apps menu -> Instance settings -> Icon admin

https://user-images.githubusercontent.com/11654230/138918402-a4999a95-72b1-4347-93c6-329b300f29b4.mov

### Changes from the last POC

* Allow adding icon packs from a spritemap file
* Allow adding icon packs from icons from other spritemaps
* Removing icon packs
* Adding icons to an icon pack from a .svg file, from a spritemap or from other spritemaps

The idea I had in mind is that we are going to have a "admin" icon pack, in this case clay, which will be the default pack. User may create different packs using icons from clay pack or adding from other places. After that user may choose a different default icon pack. Now we have `getGlobalSpritemapPath`, instead of that we may have a `getDefaultSpritemapPath` which will serve the icon pack that the user marked as default. (I wanted to include this in the POC but I rather discuss it first here).

With this changes, other use cases, like the one [Allen mentioned here ](https://github.com/liferay-frontend/liferay-portal/pull/1452#issuecomment-944634320)are possible too, since you can choose all the icons of clay but one and add a custom one for that id.

#### On Scope
Right now Icon packs admin is located in instance settings so these are scoped to the virtual instance, we may change the location and move it to Site scope. Maybe this is a good first step? Page scope feels a bit too much, at least in the first version. What do you think?

#### On Clustering
I didn't investigate anything in that scope, we are persisting the icon packs as file entries. Dani mentioned that a cluster aware repository may be created, so maybe this is feasible without changing too much the way we save those. Had to check it, though

#### On classes location
IconsUtils lives outside the portal core, this means that we cannot use it in the files from portal-web. An alternative would be to move this class into the core to make it available everywhere. Another (very ugly) option is to hardcode the path to the clay spritemap ¯\_(ツ)_/¯ 

#### On Design
I guess it is not strictly necessary in this moment, but do we have any designer assigned to the team? should we request the design to lexicon? Should I start thinking about installing figma?

#### On icon filtering
We generate the spritemaps on the flight, so maybe we want to add some query param to the spritemap url to select the icons included. Something like: `/o/icons/clay?icons=cog,add-role`

### Stuff I want to investigate (for some I think it is better to wait until we have a clearer definition)

* Suggestions from [this thread](https://github.com/liferay-frontend/liferay-portal/pull/1452#discussion_r723066563)
* svg4everybody removal (?)
* Clustering stuff

I guess that's it. Let me know what do you think :) @jbalsas @bryceosterhaus @nhpatt @dsanz 

image

### CODE REVIEW

File: modules/apps/frontend-icons/frontend-icons-admin-web/src/main/java/com/liferay/frontend/icons/admin/web/internal/portlet/action/SaveIconPackFromExistingIconsMVCActionCommand.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import java.util.Optional;
+
+import javax.portlet.ActionRequest;
+import javax.portlet.ActionResponse;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Bryce Osterhaus
```
Comment (bryceosterhaus) [inline modules/apps/frontend-icons/frontend-icons-admin-web/src/main/java/com/liferay/frontend/icons/admin/web/internal/portlet/action/SaveIconPackFromExistingIconsMVCActionCommand.java:50]: Glad I can take all credit for this 😎

**Review by bryceosterhaus — COMMENTED** [review]
From a code point of view, this looks great! I'm happy to see that it seems like you were able to update a bunch of things I had written so that it was done properly, like the portlet repository.

I'm still waiting for DXP to build so that I can test the interactions, but I am excited how this is turning out! I'll follow up after testing it.

**Comment by bryceosterhaus** [comment]
Had some issues trying to build.

```sh
     [echo] Note: Recompile with -Xlint:unchecked for details.
     [echo] /Users/bryceosterhaus/repos/liferay-portal/modules/dxp/apps/commerce/commerce-shop-by-diagram-web/build/jspc/org/apache/jsp/diagram_005ftype/svg_jsp.java:18: error: package com.liferay.frontend.icons.util does not exist
     [echo] import com.liferay.frontend.icons.util.IconsUtil;
     [echo]                                       ^
     [echo] /Users/bryceosterhaus/repos/liferay-portal/modules/dxp/apps/commerce/commerce-shop-by-diagram-web/build/jspc/org/apache/jsp/diagram_005ftype/default_jsp.java:18: error: package com.liferay.frontend.icons.util does not exist
     [echo] import com.liferay.frontend.icons.util.IconsUtil;
     [echo]                                       ^
     [echo] /Users/bryceosterhaus/repos/liferay-portal/modules/dxp/apps/commerce/commerce-shop-by-diagram-web/build/jspc/org/apache/jsp/init_jsp.java:18: error: package com.liferay.frontend.icons.util does not exist
     [echo] import com.liferay.frontend.icons.util.IconsUtil;
     [echo]                                       ^
     [echo] /Users/bryceosterhaus/repos/liferay-portal/modules/dxp/apps/commerce/commerce-shop-by-diagram-web/build/jspc/org/apache/jsp/render/list/entry/view_jsp.java:18: error: package com.liferay.frontend.icons.util does not exist
     [echo] import com.liferay.frontend.icons.util.IconsUtil;
     [echo]                                       ^
     [echo] /Users/bryceosterhaus/repos/liferay-portal/modules/dxp/apps/commerce/commerce-shop-by-diagram-web/build/jspc/org/apache/jsp/render/view_jsp.java:18: error: package com.liferay.frontend.icons.util does not exist
     [echo] import com.liferay.frontend.icons.util.IconsUtil;
     [echo]                                       ^
     [echo] /Users/bryceosterhaus/repos/liferay-portal/modules/dxp/apps/commerce/commerce-shop-by-diagram-web/build/jspc/org/apache/jsp/edit_005fcs_005fdiagram_005fsetting_jsp.java:18: error: package com.liferay.frontend.icons.util does not exist
     [echo] import com.liferay.frontend.icons.util.IconsUtil;
     [echo]                                       ^
     [echo] Note: /Users/bryceosterhaus/repos/liferay-portal/modules/dxp/apps/commerce/commerce-shop-by-diagram-web/build/jspc/org/apache/jsp/edit_005fcs_005fdiagram_005fsetting_jsp.java uses or overrides a deprecated API.
     [echo] Note: Recompile with -Xlint:deprecation for details.
```

**Comment by bryceosterhaus** [comment]
@victorg1991 would it be possible to generate that icon pack for masterclass _within_ the site initializer? Or with some deployment mechanism so that the UI steps do not have to be take?

My only concern with use case is it requires both the UI steps to be taken, and the fragments to reference a very specific name "masterclass". Doing something like `icons.getCurrentIconPack()` would be really cool, but I imagine there would have to be some "magic" behind the scenes and we would want to make sure this magic is really consistent for users so there are no "gotchas."

**Comment by ethib137** [comment]
I like the direction this is taking. I really like having `${icons.getCurrentIconPack()}`. I'm assuming this is dependent on a Site Level configuration where you would define what Icon Pack, or may be Packs would be available on the site. This is what would keep there from being too much magic, right?

**Comment by ethib137** [comment]
@victorg1991 I think another really good test case for you would be to create a site using the minium initializer, then switch the theme to classic. You'll notice that the icons for most of commerce disappear. How will this feature help to solve this situation?

File: modules/apps/frontend-icons/frontend-icons-admin-web/src/main/java/com/liferay/frontend/icons/admin/web/internal/portlet/action/SaveIconPackFromExistingIconsMVCActionCommand.java (hunk 1 above)
Comment (victorg1991) [reply modules/apps/frontend-icons/frontend-icons-admin-web/src/main/java/com/liferay/frontend/icons/admin/web/internal/portlet/action/SaveIconPackFromExistingIconsMVCActionCommand.java:50]: Hahahaha, the good old copy/paste

**Comment by victorg1991** [comment]
@bryceosterhaus 
> Had some issues trying to build.

Haven't tried with dxp enabled before, I've fixed it and rebased, should be fine now :) 

> @victorg1991 would it be possible to generate that icon pack for masterclass within the site initialiser? Or with some deployment mechanism so that the UI steps do not have to be take?
> 
> My only concern with use case is it requires both the UI steps to be taken, and the fragments to reference a very specific name "masterclass". Doing something like icons.getCurrentIconPack() would be really cool, but I imagine there would have to be some "magic" behind the scenes and we would want to make sure this magic is really consistent for users so there are no "gotchas."

Yeah, deployment mechanism doesn't play very well with clustering though,  so we probably need to enable some kind of programmatic mechanism. Also for Design packs (which is a feature aimed to serve as a light weight kind of theme) those may include icons too (this needs to be defined yet). We need to think how to integrate all of these feature.

Regarding hardcoding masterclass icon pack, I agree we shouldn't use that directly. We are thinking about adding an UI to select an icon pack for a Site, this way we just need to go for the `icons.getCurrentIconPack()`. It shouldn't be "too magical" since in the servlet we have the groupId available, so we could generate an url like `/o/icons/current` which will serve the correct icon pack depending on the site. This has the downside that the cache has to be invalidated a lot if the user is visiting several sites. Also, we can include the groupId directly in the url like `/o/icons/{groupId}` so that the url can be cached more time. What do you think about that? GroupId should be fairly easy to obtain in JSP/JS context

**Comment by victorg1991** [comment]
@ethib137 

> I like the direction this is taking. I really like having ${icons.getCurrentIconPack()}. I'm assuming this is dependent on a Site Level configuration where you would define what Icon Pack, or may be Packs would be available on the site. This is what would keep there from being too much magic, right?

Exactly that, I'd say that you can only select one, at least in the first iteration, to avoid overcomplicating it. We need to decide where to put this UI, but this need some definition since it may overlap with other features like Design packs. Do you think having only one per site may be enough for most cases?


> @victorg1991 I think another really good test case for you would be to create a site using the minium initializer, then switch the theme to classic. You'll notice that the icons for most of commerce disappear. How will this feature help to solve this situation?

Minium doesn't use fragments nor content pages so this is definitely a good test case, I'll give it a try, thanks!! :)

**Comment by ethib137** [comment]
> Do you think having only one per site may be enough for most cases?
> @victorg1991

I think it would be best if it was possible to select more than one. Behind the scenes it could create a new pack of icons, but that should be hidden from the UI. If we don't allow that users will be forced to create many extra new packs that are simply a combination `classic` icons, `commerce` icons, and `custom` icons. Here are a couple examples:

- An Intranet: Needs `Classic/Clay` icons for intranet widgets + `custom` icons for custom widgets.
- Commerce Site: Needs `Classic/Clay` + `Commerce` + maybe `custom`.

It would be better to let a site admin choose multiple icon packs that they need, rather than to have to request that someone with instance permissions creates them a new icon pack for their site.

What do you think? :)

**Comment by victorg1991** [comment]
> 
> I think it would be best if it was possible to select more than one. Behind the scenes it could create a new pack of icons, but that should be hidden from the UI. If we don't allow that users will be forced to create many extra new packs that are simply a combination `classic` icons, `commerce` icons, and `custom` icons. Here are a couple examples:
> 
> * An Intranet: Needs `Classic/Clay` icons for intranet widgets + `custom` icons for custom widgets.
> * Commerce Site: Needs `Classic/Clay` + `Commerce` + maybe `custom`.
> 
> It would be better to let a site admin choose multiple icon packs that they need, rather than to have to request that someone with instance permissions creates them a new icon pack for their site.
> 
> What do you think? :)

I wasn't thinking about serving all the icon pack selected together but I think that makes a lot of sense actually, since we are creating the spritemap on the fly this should be easy to achieve, and also useful, as you mentioned.  cc/ @dsanz

**Comment by bryceosterhaus** [comment]
> creating the spritemap on the fly this should be easy to achieve

I could imagine we might be able to group icon packs together in a single request.
`/o/icons?packs=SITE_ID,commerce,admin`

This is sort of turning into an icon specific combo loader 😂

**Comment by victorg1991** [comment]
superseded by https://github.com/liferay-frontend/liferay-portal/pull/1827/files

**Pull request closed without merging**